src/codewhisperer/client/codewhisperer.ts

```diff
@@ -39,8 +39,16 @@ export type ListCodeScanFindingsRequest = Readonly<CodeWhispererClient.ListCodeS
 export type ArtifactType = Readonly<CodeWhispererClient.ArtifactType>
 export type ArtifactMap = Readonly<CodeWhispererClient.ArtifactMap>
 export class DefaultCodeWhispererClient {
+    private credentials?: CognitoIdentityCredentials
+
     private async createSdkClient(): Promise<CodeWhispererClient> {
-        const credentials = !isCloud9() ? await getCognitoCredentials() : undefined
+        if (!this.credentials && !isCloud9()) {
+            this.credentials = await getCognitoCredentials()
+        } else if (this.credentials?.needsRefresh()) {
+            this.credentials.refresh(error => {
```

Comment:
2 changes are needed:
1. Catch all exceptions when either getting the credentials or refreshing the credentials. Throw an Error and a followup PR is needed to clean up telemetry logic when cognito API went wrong.
2.  Wait until this.credentials.refresh is finished.